Fix 2162 (at least partially) by refactoring stream writes in the fileJobStore

adamnovak opened the pull request:
I've backed out the code that I wrote to diagnose what is holding a file open and try to kill it or close the file descriptor, but this at least solves the immediate problem I was having in that issue and somehow stops the file descriptor from leaking in the first place.

It's not a general solution to misbehaving user code, which can still do bad things like start a subprocess attached to a file stream in the job store and leave it running when the Toil job exits.

adamnovak commented:
I'm duplicating some of the stack-walking code form writeFile, but I'm reluctant to pull it out into its own function because of the aformentioned stack walking it has to do, and because the rationale for why the stack walking ought to work is different in each case.

cgcloud-jenkins commented:
Build finished. 18 tests run, 0 skipped, 1 failed.

cgcloud-jenkins commented:
Build finished. 880 tests run, 33 skipped, 4 failed.

DailyDreaming commented on src/toil/jobStores/fileJobStore.py:
```diff
@@ -280,27 +285,47 @@ def _supportsUrl(cls, url, export=False):
-            sourceFunctionName = "x"
+            sourceFunctionName = "UNKNOWNJOB"
             # make sure the function name fetched has no spaces or oddities
+        finally:
```
This looks like a no-op.

DailyDreaming commented on src/toil/jobStores/fileJobStore.py:
```diff
@@ -280,27 +285,47 @@ def _supportsUrl(cls, url, export=False):
     def writeFileStream(self, jobStoreID=None):
-        fd, absPath = self._getTempFile(jobStoreID)
+        # Record the name of the job/function writing the file in the file name
+        try:
```
This looks like duplicated code from the above and should probably be its own function.

adamnovak commented:
OK @DailyDreaming, this should address your comments.

I'm not sure what went wrong with that WDL test. I'll see if I can get it to misbehave for me and maybe say more than that it is launching a job and then that it has timed out.

adamnovak commented:
I wrangled a Docker machine, and that test passed in 350.27 seconds when I ran it with:
```
make test tests=src/toil/test/wdl/toilwdlTest.py::ToilWdlIntegrationTest::testENCODE
```

Maybe something I changed to address the review comments also fixed the test? Or maybe something important is different between Jenkins and my environment.

cgcloud-jenkins commented:
Build finished. 18 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 880 tests run, 33 skipped, 0 failed.

DailyDreaming commented:
@adamnovak Strange.  Well, tests are passing now and it looks good to me.  Merging.
